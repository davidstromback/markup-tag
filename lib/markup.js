import { parse } from "./parse.js";
import { interpolate } from "./interpolate.js";
import { compile, tokenizer, memo } from "fsm-tokenizer";
export const schema = compile({
    initialState: "text",
    tokens: {
        closeMarker: {},
        equals: {},
        escape: {},
        name: { value: true },
        quote: {},
        tagCloser: {},
        tagOpener: {},
        text: { value: true },
        value: { value: true, padding: true },
    },
    states: {
        text: {
            token: "text",
            rules: [
                [/\\/, "before", "textEscape", true],
                [/</, "before", "tagOpener", true],
                [/[\s\S]/, "after", "text", false],
            ],
        },
        textEscape: {
            token: "escape",
            rules: [[/\\/, "after", "textEscaped", true]],
        },
        textEscaped: {
            token: "text",
            rules: [[/[\s\S]/, "after", "text", false]],
        },
        tagOpener: {
            token: "tagOpener",
            rules: [[/</, "after", "nameOrClose", true]],
        },
        closeMarker: {
            token: "closeMarker",
            rules: [[/\//, "after", "nameOrClose", true]],
        },
        nameOrClose: {
            token: "name",
            rules: [
                [/\s/, "after", "nameOrClose", false],
                [/[a-zA-Z]/, "before", "name", true],
                [/\//, "before", "closeMarker", true],
                [/>/, "before", "tagCloser", true],
            ],
        },
        name: {
            token: "name",
            rules: [
                [/[a-zA-Z0-9]/, "after", "name", false],
                [/\s/, "before", "nameOrClose", true],
                [/>/, "before", "tagCloser", true],
                [/\//, "before", "closeMarker", true],
                [/=/, "before", "equalsOrName", true],
            ],
        },
        equalsOrName: {
            token: "equals",
            rules: [
                [/\s/, "after", "equalsOrName", false],
                [/=/, "after", "openingQuoteOrName", true],
                [/\//, "before", "closeMarker", true],
                [/>/, "before", "tagCloser", true],
                [/[a-zA-Z]/, "before", "nameOrClose", true],
            ],
        },
        openingQuoteOrName: {
            token: "quote",
            rules: [
                [/\s/, "after", "openingQuoteOrName", false],
                [/"/, "after", "value", true],
                [/\//, "before", "closeMarker", true],
                [/>/, "before", "tagCloser", true],
                [/[a-zA-Z]/, "before", "nameOrClose", true],
            ],
        },
        value: {
            token: "value",
            rules: [
                [/\\/, "before", "valueEscape", true],
                [/"/, "before", "closingQuote", true],
                [/[\s\S]/, "after", "value", false],
            ],
        },
        valueEscape: {
            token: "escape",
            rules: [[/\\/, "after", "valueEscaped", true]],
        },
        valueEscaped: {
            token: "value",
            rules: [[/[\s\S]/, "after", "value", false]],
        },
        closingQuote: {
            token: "quote",
            rules: [[/"/, "after", "nameOrClose", true]],
        },
        tagCloser: {
            token: "tagCloser",
            rules: [[/>/, "after", "text", true]],
        },
    },
});
const tokenize = memo(tokenizer(schema, (type, value) => ({ type, value })));
export function markup(strings, ...values) {
    return parse(interpolate(tokenize, strings, values));
}
