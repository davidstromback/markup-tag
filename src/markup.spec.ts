/* eslint-disable no-template-curly-in-string */

import { markup } from "./markup.js";

describe("markup", () => {
  describe("matches snapshots", () => {
    test('<foo foo="bar" />', () => {
      expect(markup`<foo foo="bar" />`).toMatchSnapshot();
    });

    const MyComponent = () => "foo";

    const todos = ["buy milk", "git gud"];

    test("<foo attribute=${'value'} />", () => {
      expect(
        markup`
            <${"tagName"} attribute=${"value"} ${{ foo: "bar" }}>
                <div>
                child0
                ${["child1", "child2"]}
                child3
                <${MyComponent} foo="">
                    Test
                    ${todos}
                    ${todos.map((todo) => markup`<div>${todo}</div>`)}
                </${MyComponent}>
                <div />
                </div>
            </${"tagName"}>
        `
      ).toMatchSnapshot();
    });
  });
});
