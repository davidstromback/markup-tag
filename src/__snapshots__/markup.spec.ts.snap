// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markup matches snapshots <foo attribute=\${'value'} /> 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          "child0",
          "child1",
          "child2",
          "child3",
          Object {
            "children": Array [
              "Test",
              "buy milk",
              "git gud",
              Object {
                "children": Array [
                  "buy milk",
                ],
                "props": Object {},
                "type": "div",
              },
              Object {
                "children": Array [
                  "git gud",
                ],
                "props": Object {},
                "type": "div",
              },
            ],
            "props": Object {
              "foo": "",
            },
            "type": [Function],
          },
          Object {
            "children": Array [],
            "props": Object {},
            "type": "div",
          },
        ],
        "props": Object {},
        "type": "div",
      },
    ],
    "props": Object {
      "attribute": "value",
      "foo": "bar",
    },
    "type": "tagName",
  },
]
`;

exports[`markup matches snapshots <foo foo="bar" /> 1`] = `
Array [
  Object {
    "children": Array [],
    "props": Object {
      "foo": "bar",
    },
    "type": "foo",
  },
]
`;
